paasta_tools/utils.py

```diff
@@ -2759,8 +2763,63 @@ def get_services_for_cluster(
     return instance_list
 
 
-def parse_yaml_file(yaml_file: str) -> Any:
-    return yaml.safe_load(open(yaml_file))
+def load_service_instance_configs(
+    service: str, instance_type: str, cluster: str, soa_dir: str = DEFAULT_SOA_DIR,
+) -> Dict[str, Dict[str, Any]]:
+    conf_file = f"{instance_type}-{cluster}"
+    user_configs = service_configuration_lib.read_extra_service_information(
+        service, conf_file, soa_dir=soa_dir
+    )
+    user_configs = filter_templates_from_config(user_configs)
+    auto_configs = load_service_instance_auto_configs(
+        service, instance_type, cluster, soa_dir
+    )
+    merged = {}
+    for instance_name, user_config in user_configs.items():
+        auto_config = auto_configs.get(instance_name, {})
+        merged[instance_name] = deep_merge_dictionaries(
+            overrides=user_config, defaults=auto_config,
+        )
+    return merged
+
+
+def load_service_instance_config(
+    service: str,
+    instance: str,
+    instance_type: str,
+    cluster: str,
+    soa_dir: str = DEFAULT_SOA_DIR,
+) -> Dict[str, Any]:
+    if instance.startswith("_"):
+        raise InvalidJobNameError(
+            f"Unable to load {instance_type} config for {service}.{instance} as instance name starts with '_'"
+        )
+    conf_file = f"{instance_type}-{cluster}"
+    user_config = service_configuration_lib.read_extra_service_information(
+        service, conf_file, soa_dir=soa_dir
+    ).get(instance)
+    if user_config is None:
+        raise NoConfigurationForServiceError(
+            f"{instance} not found in config file {soa_dir}/{service}/{conf_file}.yaml."
+        )
+
+    auto_config = load_service_instance_auto_configs(
+        service, instance_type, cluster, soa_dir
+    ).get(instance, {})
+    return deep_merge_dictionaries(overrides=user_config, defaults=auto_config,)
+
+
+def load_service_instance_auto_configs(
+    service: str, instance_type: str, cluster: str, soa_dir: str = DEFAULT_SOA_DIR,
+) -> Dict[str, Dict[str, Any]]:
+    enabled_types = load_system_paasta_config().get_auto_config_instance_types_enabled()
```

Comment:
As designed, this could be a little expensive, loading this file on every config. Can you check to make sure this won't be a big performance hit? (not sure if there is an easy benchmark, `paasta validate` maybe?)

Reply:
Sure, worst case it doubles the config loading time (if we keep it as YAML). It will be less in practice because the auto/ files will contain less config than the regular ones.

I simulated that by copying all the marathon* files in yelp-main into the auto/ directory. With this change, `paasta validate -s yelp-main` takes 37.744s, without 29.212s. 

I also looked at how long it takes to load all marathon instances. Again, I copied the marathon files into auto/ for every service to simulate the worst case:
```
service_configuration_lib.disable_yaml_cache()

for service in list_services(soadir):
    for cluster in clusters:
        instances = read_service_instance_names(service, "marathon", cluster, soadir)
        for _, instance in instances:
            config = load_service_instance_config(service, instance, "marathon", cluster, soadir)
```
This takes 1m52.380s. As expected, that's just about 2x the time as it takes to load everything before this PR. There are about 7000 service instances, so that's about 8ms per service config before, 16ms after. Even if it's 2x slower, a few extra milliseconds seems fine to me.

Reply:
ty for confirming.
JSON will help if we do that too.

I was more concerned about load_system_paasta_config, but looking at that, it is cached, which is great!